config.yaml

```diff
@@ -25,3 +16,108 @@ options:
       to go into error state until the change is reverted.
     default: ""
     type: string
+  containerd_http_proxy:
+    description: |
+      Set an HTTP proxy to be used by containerd to pull images from image registries. This is
+      useful when deploying the charm on a constrained environment.
+
+      Example: "http://squid.internal:3128"
+    default: ""
+    type: string
+  containerd_https_proxy:
+    description: |
+      Set an HTTPS proxy to be used by containerd to pull images from image registries. This is
+      useful when deploying the charm on a constrained environment.
+
+      Example: "http://squid.internal:3128"
+    default: ""
+    type: string
+  containerd_no_proxy:
+    description: |
+      When configuring an HTTP/HTTPS proxy for containerd, specify the list of IP ranges to exclude.
+
+      Example: "127.0.0.1,10.0.0.0/8,192.168.0.0/16,172.16.0.0/12"
+    default: ""
+    type: string
+  automatic_certificate_reissue:
+    description: |
+      By default, MicroK8s will automatically regenerate server certificates when the IP address
+      of the host machine changes. This is problematic in scenarios where a bridge is configured
+      on the host and IP addresses on the bridge are changed frequently.
+
+      Set this flag to false if that is the case to prevent MicroK8s from continuously restarting
+      and affecting the service.
+
+      This option is ignored and does not have any effect on multi-node clusters.
+    default: true
+    type: boolean
+  extra_sans:
+    description: |
+      Comma-separated list of extra Subject Alternate Names that will be added to the kube-apiserver
+      certificates. This is useful in cases where the MicroK8s control plane is made accessible over
+      a floating IP or with DNS round-robin load-balancing.
+
+      The list may contain valid IP addresses or hostnames. If %UNIT_PUBLIC_ADDRESS% is in the list,
+      it will be replaced by Juju with the public address of the unit (e.g. the OpenStack floating
+      IP).
+
+      Note that the charm will not configure any DNS records or virtual IPs, these have to be setup
+      by the administrator separately.
+
+      Examples:
+
+      - ""                        # do nothing
+      - "10.10.10.10"             # add '10.10.10.10' as a SAN
+      - "k8s-1.lan"               # add 'k8s-1.lan' as a SAN
+      - "10.10.10.10,k8s-1.lan"   # add two SANs
+      - "%UNIT_PUBLIC_ADDRESS%"   # add the public unit address as a SAN
+    default: "%UNIT_PUBLIC_ADDRESS%"
+    type: string
+  containerd_custom_registries:
```

Comment:
Not a big fan of this free text config.

Reply:
The configuration is validated using Pydantic (see `class Registry` relevant code in src/containerd.py), and the charm gets into a block status if there is any issues, with a clear validation error describing the problem. This behaviour is covered by unit tests in `test_config_containerd_custom_registries`

Example status and logs:

```
$ juju config microk8s custom_containerd_configs='[{"key": "value"}]'

$ juju status
App       Version  Status   Scale  Charm     Channel  Rev  Exposed  Message
microk8s           blocked      3  microk8s            41  yes      failed to apply containerd_custom_registries, check logs for details

Unit          Workload  Agent  Machine  Public address  Ports      Message
microk8s/30   blocked   idle   0        172.16.100.226  16443/tcp  failed to apply containerd_custom_registries, check logs for details
microk8s/31   blocked   idle   1        172.16.100.18   16443/tcp  failed to apply containerd_custom_registries, check logs for details
microk8s/32*  blocked   idle   2        172.16.100.111  16443/tcp  failed to apply containerd_custom_registries, check logs for details


$ juju debug-log
...
unit-microk8s-32: 17:09:06 ERROR unit.microk8s/32.juju-log failed to configure containerd registries
Traceback (most recent call last):
  File "./src/charm.py", line 168, in config_containerd_registries
    registries = containerd.parse_registries(self.config["containerd_custom_registries"])
  File "/var/lib/juju/agents/unit-microk8s-32/charm/src/containerd.py", line 141, in parse_registries
    return RegistryConfigs(registries=parsed).registries
  File "/var/lib/juju/agents/unit-microk8s-32/charm/venv/pydantic/main.py", line 341, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 2 validation errors for RegistryConfigs
registries -> 0 -> url
  field required (type=value_error.missing)
registries -> 0 -> key
  extra fields not permitted (type=value_error.extra)
```